controllers/meshgateway_controller.go

```diff
@@ -39,15 +55,161 @@ type MeshGatewayReconciler struct {
+	logger := r.Log.WithValues("meshgateway", req.NamespacedName)
+
+	mgw := &servicemeshv1alpha1.MeshGateway{}
+	err := r.Get(context.TODO(), req.NamespacedName, mgw)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	logger.Info("reconciling")
+
+	reconciler, err := GetHelmReconciler(r, meshgateway.NewChartReconciler)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	icp, err := r.getRelatedIstioControlPlane(ctx, r.GetClient(), mgw, logger)
+	// do not reqeue if control plane was not found
+	if icp == nil && err == nil {
+		return ctrl.Result{
+			Requeue: false,
+		}, nil
+	}
+
+	if !isIstioVersionSupported(icp.Spec.Version) {
+		return ctrl.Result{}, nil
+	}
+
+	err = util.AddFinalizer(r.Client, mgw, meshGatewayFinalizerID)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	result, err := reconciler.Reconcile(mgw)
+	if err != nil {
+		return result, errors.WrapIf(err, "could not reconcile mesh gateway")
+	}
+
+	if result.Requeue {
+		result.RequeueAfter = 0
 
-	// your logic here
+		return result, nil
+	}
+
+	result, err = r.setGatewayAddress(ctx, r.GetClient(), mgw, logger, result)
+	if err != nil {
+		return result, errors.WrapIf(err, "could not set gateway address")
+	}
+
+	err = util.RemoveFinalizer(r.Client, mgw, istioControlPlaneFinalizerID)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return result, nil
+}
+
+func (r *MeshGatewayReconciler) GetClient() client.Client {
+	return r.Client
+}
+
+func (r *MeshGatewayReconciler) GetScheme() *runtime.Scheme {
+	return r.Scheme
+}
+
+func (r *MeshGatewayReconciler) GetLogger() logr.Logger {
+	return r.Log
+}
 
-	return ctrl.Result{}, nil
+func (r *MeshGatewayReconciler) GetName() string {
+	return "meshgateway"
 }
 
 func (r *MeshGatewayReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
-		For(&servicemeshv1alpha1.MeshGateway{}).
+		For(&servicemeshv1alpha1.MeshGateway{}, ctrlBuilder.WithPredicates(util.ObjectChangePredicate{})).
 		Complete(r)
 }
+
+func (r *MeshGatewayReconciler) getRelatedIstioControlPlane(ctx context.Context, c client.Client, mgw *servicemeshv1alpha1.MeshGateway, logger logr.Logger) (*servicemeshv1alpha1.IstioControlPlane, error) {
+	icp, err := GetRelatedIstioCR(context.Background(), r.GetClient(), client.ObjectKey{
+		Name:      mgw.GetSpec().GetIstioControlPlane().GetName(),
+		Namespace: mgw.GetSpec().GetIstioControlPlane().GetNamespace(),
+	})
+	if err != nil {
+		updateErr := components.UpdateStatus(ctx, c, mgw, components.ConvertConfigStateToReconcileStatus(servicemeshv1alpha1.ConfigState_ReconcileFailed), err.Error())
+		if updateErr != nil {
+			logger.Error(updateErr, "failed to update state")
```

Comment:
same suggestion for all similar error logs
```suggestion
			logger.Error(updateErr, "failed to update mesh gateway state")
```